include/jemalloc/internal/ehooks.h

```diff
@@ -1,4 +1,151 @@
 #ifndef JEMALLOC_INTERNAL_EHOOKS_H
 #define JEMALLOC_INTERNAL_EHOOKS_H
 
+#include "jemalloc/internal/atomic.h"
+
+extern const extent_hooks_t extent_hooks_default;
+
+typedef struct ehooks_s ehooks_t;
+struct ehooks_s {
+	/* Logically an extent_hooks_t *. */
+	atomic_p_t ptr;
+};
+
+void ehooks_init(ehooks_t *ehooks, extent_hooks_t *extent_hooks);
+
+static inline void
+ehooks_set_extent_hooks_ptr(ehooks_t *ehooks, extent_hooks_t *extent_hooks) {
+	atomic_store_p(&ehooks->ptr, extent_hooks, ATOMIC_RELEASE);
+}
+
+static inline extent_hooks_t *
+ehooks_get_extent_hooks_ptr(ehooks_t *ehooks) {
+	return (extent_hooks_t *)atomic_load_p(&ehooks->ptr, ATOMIC_ACQUIRE);
+}
+
+static inline bool
+ehooks_are_default(ehooks_t *ehooks) {
+	return ehooks_get_extent_hooks_ptr(ehooks) == &extent_hooks_default;
+}
+
+static inline bool
+ehooks_destroy_is_noop(ehooks_t *ehooks) {
+	return ehooks_get_extent_hooks_ptr(ehooks)->destroy == NULL;
+}
+
+static inline bool
+ehooks_purge_lazy_will_fail(ehooks_t *ehooks) {
+	return ehooks_get_extent_hooks_ptr(ehooks)->purge_lazy == NULL;
+}
+
+static inline bool
+ehooks_purge_forced_will_fail(ehooks_t *ehooks) {
+	return ehooks_get_extent_hooks_ptr(ehooks)->purge_forced == NULL;
+}
+
+static inline bool
+ehooks_split_will_fail(ehooks_t *ehooks) {
+	return ehooks_get_extent_hooks_ptr(ehooks)->split == NULL;
+}
+
+static inline bool
+ehooks_merge_will_fail(ehooks_t *ehooks) {
+	return ehooks_get_extent_hooks_ptr(ehooks)->merge == NULL;
+}
+
+static inline void *
+ehooks_alloc(ehooks_t *ehooks, void *new_addr, size_t size, size_t alignment,
+    bool *zero, bool *commit, unsigned arena_ind) {
+	extent_hooks_t *extent_hooks = ehooks_get_extent_hooks_ptr(ehooks);
+	return extent_hooks->alloc(extent_hooks, new_addr, size, alignment,
```

Comment:
Currently, we disallow a NULL alloc hook (i.e. this version is the behavior-neutral one). I think it arguably makes sense to allow it, and treat an attempt to invoke it as a failed/declined request (the same way that a NULL merge call is treated as the hooks refusing to merge). This could be used (for example) to create an arena that can't consume any more memory than it has currently allocated. But at least in the short term I figured we'd minimize changes here (especially ones that commit us to more guarantees while we're thinking of changes in this area).

Reply:
Makes sense. Maybe add an assertion when init the hooks.